### DISCUSSION TITLE
Object union with optional parameter

### DISCUSSION THREAD

**Discussion by Treiblesschorle** [open-policy-agent/community#46] [OPA and Rego] [answered]
Hi,
I'm trying to get the union of two objects where one of the objects may be optional. i.e. undefined (one object is coming from `data` and the other from `input` and may or may not be there). I tried using `object.union(objectA, objectB)`, however if one of the args is undefined the result is also undefined. 
Does somebody have an idea how to solve that issue?

**Comment by srenatus** [comment, marked as answer]
You could use a rule, with one body per (defined, undefined) x (defined, undefined) pair, e.g.

```rego
a = data.foo
b = input.bar

union_maybe = object.union(a, b)

union_maybe = a {
  not b
}

union_maybe = b {
  not a
}
```
👉 [Playground](https://play.openpolicyagent.org/p/e9GL3WfOts)

**Reply by Treiblesschorle** [reply]
Cool, that works. Thanks!
I tried something like this:
```
get_or_default(a, b) = res {
	not a
	res := a
} else = res {
	res := b
}
```
but it doesn't work, and I'm not sure why.

**Reply by Treiblesschorle** [reply]
`union_maybe` as a function would be handy as I'd like to use it in several places, so I don't have to hardcode the input.

**Reply by srenatus** [reply]
It wouldn't work like that function because `get_or_default(data.foo, input.bar)` wouldn't get evaluated unless both its operands were defined. That's how function calls work in OPA, I'm afraid.

We've been discussing a "defaulting dot" kind of operator, which would probably be of use, here, too: See https://github.com/open-policy-agent/opa/issues/2345 for example.

**Reply by anderseknert** [reply]
While it won't work if the elements are in the "root" data location (i.e. data.x) as that would potentially be recursive you could use the `object.get` built-in function like `object.get(data.x, "y", "my_default_value")` for some cases like this.

**Reply by Treiblesschorle** [reply]
Thanks for the answers. So my initial problem is that the two objects I want the union of do actually contain further objects, i.e. I want to do a deep merge of the two objects. As the overall structure does not change, I tried the union on each of the nested objects. So a deep object merge would also solve my problem.

**Reply by Treiblesschorle** [reply]
For now I solved it with a bunch of nested `object.get(data.x, "y", "my_default_value")` calls. However, I think adding something like mentioned here open-policy-agent/opa#2345 would be nice. Thanks for the quick help guys :)

**Reply by anderseknert** [reply]
Indeed, that would be nice!
